src/CacheManager.Core/BaseCacheManager.Keys.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using CacheManager.Core.Logging;
+
+namespace CacheManager.Core
+{
+    public sealed partial class BaseCacheManager<TCacheValue>
+    {
+        /// <inheritdoc />
+        override public IEnumerable<string> Keys(string pattern, string region)
+        {
+            CheckDisposed();
+
+            var keys = new HashSet<string>();
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] started.", region, pattern);
+            }
+
+            for (var handleIndex = 0; handleIndex < _cacheHandles.Length; handleIndex++)
+            {
+                var handle = _cacheHandles[handleIndex];
+                var handleKeys = handle.Keys(pattern ?? "*", region);
+                keys.UnionWith(handleKeys);
+            }
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] found {2} keys.", region, pattern, keys.Count);
+            }
+
+            if (region != null)
+            {
+                var after = region.Length + 1; // +1 for the :
+                return keys.Select(k => k.Substring(after));
```

Comment:
I'm not expecting every case to use "region:", but enough of them did that it seemed useful to pull the shared code up. For those that don't, they would override the function (Keys(pattern, region)) rather than just implementing AllKeys().

The substring is so that it passes back the plan keys (without region). 
It was either do it at this level, or at the remove check to see if it started with the region or if it needed to be prepended. This seemed like the less messy solution.